typing/env.ml

```diff
@@ -2901,32 +2916,55 @@ let lookup_ident_module (type a) (load : a load) ~errors ~use ~loc s env =
 
 let lock_mode ~errors ~loc env id vmode locks =
   List.fold_left
-    (fun vmode lock ->
+    (fun (vmode, reason) lock ->
       match lock with
-      | Region_lock -> Value_mode.local_to_regional vmode
-      | Lock {mode; escaping_context} ->
-        begin
-          match Value_mode.submode vmode (Value_mode.of_alloc mode) with
-          | Ok () -> vmode
+      | Region_lock -> (Mode.Value.local_to_regional vmode, reason)
+      | Locality_lock {mode; escaping_context} -> begin
+          match
+            Mode.Regionality.submode
+              (Mode.Value.locality vmode)
+              (Mode.Regionality.of_locality mode)
+          with
+          | Ok () -> (vmode, reason)
           | Error _ ->
               may_lookup_error errors loc env
                 (Local_value_used_in_closure (id, escaping_context))
         end
-      | Exclave_lock ->
-        match Value_mode.submode vmode Value_mode.regional with
-        | Ok () -> Value_mode.regional_to_local vmode
-        | Error _ ->
-          may_lookup_error errors loc env
-            (Local_value_used_in_exclave id);
-    )
-    vmode locks
+      | Linearity_lock {mode;shared_context} -> begin
+          (match Mode.Linearity.submode (Mode.Value.linearity vmode) mode with
+          | Error _ ->
+              may_lookup_error errors loc env
+              (Once_value_used_in (id, shared_context))
+          | Ok () -> ()
+          );
+          (* outside unique values can be accessed inside a loop but only at shared *)
+          let min_uniq =
+            Mode.Uniqueness.join
+              [ Mode.Value.uniqueness vmode;
+                Mode.Linearity.to_dual mode ]
+          in
+          let vmode = Mode.Value.with_uniqueness min_uniq vmode in
+          vmode, Some shared_context
+        end
```

Comment:
I like the generality of this, but also feel it would be nicer if implemented with the new solver - can I do it there instead? (considering that the code is already correct, and the new solver PR is not far from merging)

Reply:
Yeah, that's fine. You should still make the change to have a shared `Closure_lock` in this branch though.